### PR TITLE
fix: certification to private on private timeline

### LINKED ISSUE #40403: Opening a cert on a private profile redirects to the homepage
If I set my profile to private and try to open a cert from the settings page or from the timeline on my portfolio, I get a message that I need to set my profile to public to view a cert and get redirected to the home page. Seems like a user should get redirected to the page they came from.

Or possibly, remove the flash message and redirect in favor of something like we do for the portfolio pages - where I can open a cert with my profile private, but it says something like "Your profile is private, only you can see this certificate. Make your profile public in order for others to see it."

### LINKED ISSUE #42621: Private timeline hides cert solutions
If you set the timeline to private and open one of your certifications, all the solutions will be hidden:
Screen Shot 2021-06-24 at 8 03 29 AM

### PR DESCRIPTION
**Pull request by ShaunSHamilton**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #42621 

**Changes**

- Viewing a certification when `showTimeline: false` yields flash explaining this is not possible
- <s>Upon api-response with flash, redirect goes to last location</s>

### CODE REVIEW

File: client/src/redux/show-cert-saga.js [hunk 1]
```diff
@@ -13,7 +12,8 @@ function* getShowCertSaga({ payload: { username, certSlug } }) {
       for (let i = 0; i < messages.length; i++) {
         yield put(createFlashMessage(messages[i]));
       }
-      yield call(navigate, '/');
+      // eslint-disable-next-line no-restricted-globals
+      history.go(-1);
```
Comment (ShaunSHamilton) [inline client/src/redux/show-cert-saga.js:16]: To be honest, I have no idea if this is a suitable option, but did not find another way without using `redirectQueryParams` which would get/look messy.

Ideas are welcome

Comment (nhcarrigan) [reply client/src/redux/show-cert-saga.js:16]: Honestly I'd leave this as is until we have a resolution for #45563

**Review by ArielLeslie — DISMISSED** [review]

**Review by raisedadead — DISMISSED** [review]
LGTM. @nhcarrigan resolve and merge when happy.

File: client/src/redux/show-cert-saga.js (hunk 1 above)
Comment (ShaunSHamilton) [reply client/src/redux/show-cert-saga.js:16]: @nhcarrigan To clarify: Are you suggesting blocking this PR until `45563` is resolved? Or, going ahead with this as is?

Comment (ojeytonwilliams) [inline client/src/redux/show-cert-saga.js:16]:
```suggestion
      yield call(navigate, -1);
```
This is a little more consistent with how we navigate everywhere else in the app.  Also, reach-router should be more a11y friendly than calling the web apis directly.

**Review by ojeytonwilliams — DISMISSED** [review]
If we're going ahead with the PR with the navigation as is, I think we should use `navigate`, not history.

Otherwise, this LGTM.

File: client/src/redux/show-cert-saga.js (hunk 1 above)
Comment (ShaunSHamilton) [reply client/src/redux/show-cert-saga.js:16]: This does not work well for non-signed-in Campers. That is, it causes the flash to not show.

Comment (nhcarrigan) [reply client/src/redux/show-cert-saga.js:16]: I'm suggesting to leave this as `yield call(navigate, '/');` until we've implemented the FAQ page.

**Review by moT01 — COMMENTED** [review]
I gave this some testing...

If I go to the cert directly, e.g. enter `http://localhost:8000/certification/developmentuser/responsive-web-design` in the address bar - I get taken back to whatever site I came from, maybe google. So I think that leaving it as navigate to `/` might be the way to go. It might be better anyway - I dunno, I don't really like sending users back to some unknown site.

If I open the cert using the show cert button on the superblock page, I get like a page refresh or something and taken back to `http://localhost:8000/learn/responsive-web-design/#responsive-web-design-projects` and don’t get the flash message.

Works as expected when I open the cert from the settings page.

File: client/src/redux/show-cert-saga.js (hunk 1 above)
Comment (nhcarrigan) [inline client/src/redux/show-cert-saga.js:16]:
```suggestion
      yield call(navigate, '/');
```

Put this back for now so we can unblock this and move forward.

**Review by naomi-lgbt — DISMISSED** [review]

File: client/src/redux/show-cert-saga.js [hunk 2]
```diff
@@ -1,4 +1,3 @@
-import { navigate } from 'gatsby';
```
Comment (nhcarrigan) [inline client/src/redux/show-cert-saga.js:1]: gotta put this back

__
[sema-logo](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix&nbsp; | &nbsp;**Tags:** Inefficient

**Comment by ShaunSHamilton** [comment]
Opening this as is to close the issue. This at least brings the behaviour in line with the rest of the required settings.

After this, we can work on a better solution to navigate Campers based on where they previously were.

**Comment by raisedadead** [comment]
> Edit: If Oliver dares to bring up the mixture of client+api changes in one PR, I will 👊 👋 🤝 🐟

I usually second you, but I will have his back this time. Bring it on. 

> Seriously, I think we need to figure out a way to do client+api changes/deployments properly as a priority.

Yeah - in an ideal world, we should be able. We could dynamically use feature flags that we can turn on from the backend.

The point of the "current way of doing things" in multiple PRs is to manage expectations correctly. The client and the API are two separate entities. They are no different than the apps being in two different repositories (which was the case a few years ago). 

It's just easier logistically to know that an endpoint the client expects is deployed ahead of time, hence the current flow.

So, we could do dynamic feature flags; still, the client will not have access to the API endpoint it expects right away.

But one thing is set in stone though we are not going to marry the API and the Client together.

**Comment by raisedadead** [comment]
The simplest way to achieve this parity would be to test both (cilent+api) changes locally, then "comment" out parts of the client that can be uncommented in a subsequent deployment.

**Comment by ojeytonwilliams** [comment]
As I understand things, the best we can do in principle is three deployments (where I'm using deployment to mean "change something in production", that could be an environment variable or code, doesn't matter):

1. Expand api so that it can handle old and new requests
2. Switch client to make new requests
3. Contract api so that it can only handle new requests

I like @raisedadead's feature flag approach, though.  We could have a flag "MIGRATION_STAGE" that's either "api" or "client".  Then the logic would be

api:
```
if (MIGRATION_STAGE=='api') // expanded
if (MIGRATION_STAGE=='client') // expanded
else // contracted
```


client:
```
if (MIGRATION_STAGE=='api') // old requests
if (MIGRATION_STAGE=='client') // new requests
else // new requests
```

Then we can test all three stages by changing the variable.

> I will fist_oncoming wave handshake fish

Hand me a fish?

**Comment by raisedadead** [comment]
> Hand me a fish?

Send us your address.

**Comment by ShaunSHamilton** [comment]
For this PR, can we not deploy as is? All it will cause is maybe an early return from a request, or the current behaviour. There are no new endpoints, or removed endpoints.

---

> > I will fist_oncoming wave handshake fish
>
> Hand me a fish?

I was threatening your fish.

**Review by raisedadead — COMMENTED** [review]
I think the `if (!user.name) {` check and be moved right before the `if (!showCerts) {` - other than that all this should be good

**Review by moT01 — APPROVED** [review]
LGTM 🎉

**Review by raisedadead — APPROVED** [review]

**Pull request merged**